modin/pandas/test/utils.py

```diff
@@ -573,7 +584,81 @@ def assert_empty_frame_equal(df1, df2):
         assert False, f"Empty frames have different types: {type(df1)} != {type(df2)}"
 
 
-def df_equals(df1, df2):
+def assert_all_act_same(condition, *objs):
+    """
+    Assert that all of the objs give the same boolean result for the passed condition (either all True or all False).
+
+    Parameters
+    ----------
+    condition : callable(obj) -> bool
+        Condition to run on the passed objects.
+    *objs :
+        Objects to pass to the condition.
+    """
+    results = [condition(obj) for obj in objs]
+    if len(results) < 2:
+        return
+
+    assert all(results[0] == res for res in results[1:])
+
+
+def _maybe_cast_to_pandas_dtype(dtype):
+    """Cast passed `dtype` to according pandas dtype if needed for the sake of equality comparison."""
+    # If we're running in a cloud mode then all the numpy types are substituted with a proxy net-reference,
+    # Such dtypes won't pass equality check with pandas dtypes thus manually converting them to a pure numpy
+    if "netref" in str(type(dtype)):
+        return np.dtype(dtype.name)
+    return dtype
+
+
+def assert_dtypes_equal(df1, df2):
+    """
+    Assert that the two passed DataFrame/Series objects have equal dtypes.
+
+    The function doesn't require that the dtypes are identical, it has the following reliefs:
+        1. The dtypes are not required to be in the same order
+           (e.g. {"col1": int, "col2": float} == {"col2": float, "col1": int})
+        2. The dtypes are only required to be in the same class
+           (e.g. both numerical, both categorical, etc...)
+
+    Parameters
+    ----------
+    df1 : DataFrame or Series
+    df2 : DataFrame or Series
+    """
+    if not isinstance(
+        df1, (pandas.Series, pd.Series, pandas.DataFrame, pd.DataFrame)
+    ) or not isinstance(
+        df2, (pandas.Series, pd.Series, pandas.DataFrame, pd.DataFrame)
+    ):
+        return
+
+    if isinstance(df1.dtypes, (pandas.Series, pd.Series)):
+        dtypes1 = df1.dtypes
+        dtypes2 = df2.dtypes
+    else:
+        # Case when `dtypes` is a scalar
+        dtypes1 = pandas.Series({"col": df1.dtypes})
+        dtypes2 = pandas.Series({"col": df2.dtypes})
+
+    # Don't require for dtypes to be in the same order
+    assert len(dtypes1.index.difference(dtypes2.index)) == 0
+    assert len(dtypes1) == len(dtypes2)
+
+    for col in dtypes1.keys():
+        type1, type2 = map(_maybe_cast_to_pandas_dtype, (dtypes1[col], dtypes2[col]))
+        assert_all_act_same(is_numeric_dtype, type1, type2)
+        assert_all_act_same(
+            lambda obj: is_object_dtype(obj) or is_string_dtype(obj), type1, type2
+        )
+        assert_all_act_same(is_bool_dtype, type1, type2)
+        assert_all_act_same(is_categorical_dtype, type1, type2)
+        assert_all_act_same(is_datetime64_any_dtype, type1, type2)
+        assert_all_act_same(is_timedelta64_dtype, type1, type2)
+        assert_all_act_same(is_period_dtype, type1, type2)
```

Comment:
If a type cannot belong to two classes at once, maybe it's worth stopping at the check of the class to which the types belong?

Reply:
done